backend/src/errors.rs

```diff
@@ -3,89 +3,94 @@ use axum::{
     response::{IntoResponse, Json, Response},
 };
 use serde_json::json;
-use sqlx::Error as SqlxError;
-use std::fmt;
+use sqlx::{error::DatabaseError, Error as SqlxError};
 
 pub enum CustomError {
-    NotFound(String),
-    BadRequest,
-    DatabaseError(SqlxError),
+    BadRequest(String),
     InternalServerError,
 }
 
+/// Converts `sqlx::Error` into `CustomError`.
+///
+/// Enables using the `?` operator in functions returning `Result<T, CustomError>`
+/// when interacting with functions that return `Result<T, sqlx::Error>`.
+///
+/// # Example
+///
+/// ```ignore
+/// fn my_function() -> Result<(), CustomError> {
+///     another_function()?;  // This function returns `Result<(), sqlx::Error>`
+///     Ok(())
+/// }
+/// ```
 impl From<SqlxError> for CustomError {
     fn from(err: SqlxError) -> Self {
         match err {
-            SqlxError::Database(e) => CustomError::DatabaseError(sqlx::Error::Database(e)),
+            SqlxError::Database(db_error) => {
+                let pg_error = db_error.downcast::<sqlx::postgres::PgDatabaseError>();
+                if pg_error.is_unique_violation() {
+                    return CustomError::BadRequest(format!(
+                        "Unique violation: {}",
+                        pg_error.message()
+                    ));
+                } else if pg_error.is_check_violation() {
+                    return CustomError::BadRequest(format!(
+                        "Check violation: {}",
+                        pg_error.message()
+                    ));
+                } else if pg_error.is_foreign_key_violation() {
+                    return CustomError::BadRequest(format!(
+                        "Foreign violation: {}",
+                        pg_error.message()
+                    ));
+                } else if pg_error.is_unique_violation() {
+                    return CustomError::BadRequest(format!(
+                        "Duplicate key: {}",
+                        pg_error.message()
+                    ));
+                }
+                CustomError::BadRequest("A database error occurred".to_string())
```

Comment:
Hm? Should the catchall be a 400?

Reply:
BadRequest is a 400, no?

Reply:
Because the From:: block converts `BadRequest` to `http::status::BAD_REQUEST`

Reply:
I know that.  I mean if we get any other DB error, is it probably a 500? Idk what else would constitute a 400

Reply:
Ah I see, you're saying it should be a 500. That makes sense